src/plugins/tonegen/dbusif.c

```diff
@@ -0,0 +1,331 @@
+        goto failed;
+    }
+
+    dbus_error_init(&err);
+    ret = dbus_bus_request_name(dbusif->conn, service, DBUS_NAME_FLAG_REPLACE_EXISTING,
+                                &err);
+    if (ret < 0) {
+        N_ERROR(LOG_CAT "%s(): request name failed: %s", __FUNCTION__, err.message);
+        dbus_error_free(&err);
+        goto failed;
+    }
+    if (ret != DBUS_REQUEST_NAME_REPLY_PRIMARY_OWNER) {
+        N_ERROR(LOG_CAT "%s(): not a primary owner", __FUNCTION__);
+        goto failed;
+    }
+
+    dbusif->hash   = g_hash_table_new_full(g_str_hash, g_str_equal,
+                                           destroy_key, NULL);
+    N_DEBUG(LOG_CAT "D-Bus legacy API setup OK");
+
+    return dbusif;
+
+ failed:
+    if (dbusif != NULL) {
+        if (dbusif->conn)
+            dbus_connection_unref(dbusif->conn);
+        free(dbusif);
+    }
+
+    return NULL;
+}
+
+void dbusif_destroy(struct dbusif *dbusif)
+{
+    if (dbusif) {
+
+        if (dbusif->hash != NULL)
+            g_hash_table_destroy(dbusif->hash);
+
+        free(dbusif);
+    }
+}
+
+
+int dbusif_register_input_method(struct tonegend *tonegend,
+                                 char *intf, char *memb, char *sig,
+                                 int (*method)(DBusMessage*, struct tonegend*))
+{
+    struct dbusif *dbusif = tonegend->dbus_ctx;
+    gchar         *key;
+
+    if (!memb || !sig || !method) {
+        N_ERROR(LOG_CAT "%s(): Called with invalid argument(s)", __FUNCTION__);
+        errno = EINVAL;
+        return -1;
+    }
+
+    if (intf == NULL)
+        intf = service;
+
+    key = create_key((gchar *)memb, (gchar *)sig, (gchar *)intf);
+    g_hash_table_insert(dbusif->hash, key, (gpointer)method);
+
+    return 0;
+}
+
+
+int dbusif_unregister_input_method(struct tonegend *tonegend, char *intf,
+                                   char *memb, char *sign)
+{
+    (void)tonegend;
+    (void)intf;
+    (void)memb;
+    (void)sign;
+
+    return 0;
+}
+
+int dbusif_send_signal(struct tonegend *tonegend, char *intf, char *name,
+                       int first_arg_type, ...)
+{
+    struct dbusif *dbusif = tonegend->dbus_ctx;
+    DBusMessage   *msg;
+    va_list        ap;
+    int            success;
+    
+    do { /* not a loop */
+        success = FALSE;
+
+        if (name == NULL) {
+            N_ERROR(LOG_CAT "%s(): Called with invalid argument", __FUNCTION__);
+            errno   = EINVAL;
+            break;
+        }
+
+        if (intf == NULL)
+            intf = service;
+
+        if ((msg = dbus_message_new_signal(path, intf, name)) == NULL) {
+            errno = ENOMEM;
+            break;
+        }
+
+        va_start(ap, first_arg_type);
+
+        if (dbus_message_append_args_valist(msg, first_arg_type, ap)) {
+            success = dbus_connection_send(dbusif->conn, msg, NULL);
+        }
+
+        va_end(ap);
+
+        dbus_message_unref(msg);
+
+    } while(FALSE);
```

Comment:
As long as there's nothing else coming here after the do..while, the `break;` could be replaced by a `return ...;` and the do..while removed.